File: mobile/lib/models/conversation.dart
import 'user.dart';

class Conversation {
  final String id;
  final String memberId;
  final String sellerId;
  final String? listingId;
  final String lastMessageAt;
  final String createdAt;
  final PublicProfile? otherUser;
  final int unreadCount;

  Conversation({
    required this.id,
    required this.memberId,
    required this.sellerId,
    this.listingId,
    required this.lastMessageAt,
    required this.createdAt,
    this.otherUser,
    this.unreadCount = 0,
  });

  factory Conversation.fromJson(Map<String, dynamic> json) => Conversation(
        id: json['id'] as String,
        memberId: json['member_id'] as String,
        sellerId: json['seller_id'] as String,
        listingId: json['listing_id'] as String?,
        lastMessageAt: json['last_message_at'] as String,
        createdAt: json['created_at'] as String,
        otherUser: json['other_user'] != null
            ? PublicProfile.fromJson(json['other_user'] as Map<String, dynamic>)
            : null,
        unreadCount: json['unread_count'] as int? ?? 0,
      );
}

class Message {
  final String id;
  final String conversationId;
  final String senderId;
  final String content;
  final String type;
  final String? readAt;
  final String? replyToId;
  final ReplyMessage? replyTo;
  final List<MessageReaction> reactions;
  final String createdAt;

  Message({
    required this.id,
    required this.conversationId,
    required this.senderId,
    required this.content,
    this.type = 'text',
    this.readAt,
    this.replyToId,
    this.replyTo,
    this.reactions = const [],
    required this.createdAt,
  });

  factory Message.fromJson(Map<String, dynamic> json) => Message(
        id: json['id'] as String,
        conversationId: json['conversation_id'] as String,
        senderId: json['sender_id'] as String,
        content: json['content'] as String,
        type: json['type'] as String? ?? 'text',
        readAt: json['read_at'] as String?,
        replyToId: json['reply_to_id'] as String?,
        replyTo: json['reply_to'] != null
            ? ReplyMessage.fromJson(json['reply_to'] as Map<String, dynamic>)
            : null,
        reactions: (json['reactions'] as List<dynamic>?)
                ?.map((r) => MessageReaction.fromJson(r as Map<String, dynamic>))
                .toList() ??
            [],
        createdAt: json['created_at'] as String,
      );

  bool get isRead => readAt != null;
}

class ReplyMessage {
  final String id;
  final String senderId;
  final String content;
  final String type;

  ReplyMessage({
    required this.id,
    required this.senderId,
    required this.content,
    required this.type,
  });

  factory ReplyMessage.fromJson(Map<String, dynamic> json) => ReplyMessage(
        id: json['id'] as String,
        senderId: json['sender_id'] as String,
        content: json['content'] as String,
        type: json['type'] as String? ?? 'text',
      );
}

class MessageReaction {
  final String id;
  final String messageId;
  final String userId;
  final String emoji;

  MessageReaction({
    required this.id,
    required this.messageId,
    required this.userId,
    required this.emoji,
  });

  factory MessageReaction.fromJson(Map<String, dynamic> json) => MessageReaction(
        id: json['id'] as String,
        messageId: json['message_id'] as String,
        userId: json['user_id'] as String,
        emoji: json['emoji'] as String,
      );
}

File: backend/internal/model/conversation.go
package model

import "time"

type Conversation struct {
	ID            string     `json:"id"`
	MemberID      string     `json:"member_id"`
	SellerID      string     `json:"seller_id"`
	ListingID     *string    `json:"listing_id,omitempty"`
	LastMessageAt time.Time  `json:"last_message_at"`
	CreatedAt     time.Time  `json:"created_at"`
	// Joined fields
	OtherUser     *PublicProfile `json:"other_user,omitempty"`
	LastMessage   *Message       `json:"last_message,omitempty"`
	UnreadCount   int            `json:"unread_count"`
}

type Message struct {
	ID             string            `json:"id"`
	ConversationID string            `json:"conversation_id"`
	SenderID       string            `json:"sender_id"`
	Content        string            `json:"content"`
	Type           string            `json:"type"`
	ReadAt         *time.Time        `json:"read_at,omitempty"`
	ReplyToID      *string           `json:"reply_to_id,omitempty"`
	ReplyTo        *ReplyMessage     `json:"reply_to,omitempty"`
	Reactions      []MessageReaction `json:"reactions,omitempty"`
	CreatedAt      time.Time         `json:"created_at"`
}

type ReplyMessage struct {
	ID       string `json:"id"`
	SenderID string `json:"sender_id"`
	Content  string `json:"content"`
	Type     string `json:"type"`
}

type MessageReaction struct {
	ID        string    `json:"id"`
	MessageID string    `json:"message_id"`
	UserID    string    `json:"user_id"`
	Emoji     string    `json:"emoji"`
	CreatedAt time.Time `json:"created_at"`
}

type CreateConversationRequest struct {
	SellerID  string  `json:"seller_id" binding:"required"`
	ListingID *string `json:"listing_id"`
}

type SendMessageRequest struct {
	Content   string  `json:"content" binding:"required,max=5000"`
	Type      string  `json:"type" binding:"omitempty,max=50"`
	ReplyToID *string `json:"reply_to_id"`
}

type ToggleReactionRequest struct {
	Emoji string `json:"emoji" binding:"required,max=10"`
}
